### ISSUE TITLE
Value conversion causes exception when using FirstOrDefault()

### ISSUE THREAD

**Issue by lakeman** [aspnet/EntityFrameworkCore#17857]
I'm trying to return details of a parent row & optional matching child row. Where the child row has an enum, saved to the database as a string.

Repro;
``` C#
class Context : DbContext
{
	public DbSet<Parent> Parent { get; set; }
	public DbSet<Child> Child { get; set; }

	protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
	{
		optionsBuilder
			.UseSqlServer(@"Server=(localdb)\mssqllocaldb;Database=_ModelApp;Trusted_Connection=True;Connect Timeout=5;ConnectRetryCount=0");
	}

	protected override void OnModelCreating(ModelBuilder modelBuilder)
	{
		var child = modelBuilder.Entity<Child>();

		child
			.Property(e => e.ChildType)
			.HasConversion<string>();
		
		child.HasOne(e => e.Parent)
			.WithMany(e => e.Child)
			.HasForeignKey(c => c.ParentId);
	}
}

class Parent
{
	public int Id { get; set; }
	[StringLength(50)]
	public string Name { get; set; }
	public virtual ICollection<Child> Child { get; set; }
}

enum ChildType
{
	TypeA,
	TypeB
}

class Child
{
	public int Id { get; set; }
	public int ParentId { get; set; }
	public bool Primary { get; set; }
	public ChildType ChildType { get; set; }
	public virtual Parent Parent { get; set; }
}

class View
{
	public string ParentName { get; set; }
	public ChildType ChildType { get; set; }
}

using (var db = new Context())
{
	var results = db.Parent
		.Select(p => new View
		{
			ParentName = p.Name,
			ChildType = p.Child.FirstOrDefault(c => c.Primary).ChildType
		})
		.ToList();
}
```

Exception; 

```
System.InvalidCastException
  HResult=0x80004002
  Message=Unable to cast object of type 'System.String' to type 'System.Int32'.
  Source=System.Data.SqlClient
  StackTrace:
   at System.Data.SqlClient.SqlBuffer.get_Int32()
   at Microsoft.EntityFrameworkCore.Storage.Internal.TypedRelationalValueBufferFactory.Create(DbDataReader dataReader)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.BufferlessMoveNext(DbContext _, Boolean buffer)
   at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.MoveNext()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext()
   at System.Collections.Generic.List`1.AddEnumerable(IEnumerable`1 enumerable)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Repro.Program.Main(String[] args) in ...
```
Single(), SingleOrDefault() or First() don't hit above error.

EF Core version: 2.2.6
Database provider: SqlServer
Target framework: .NET Core 2.2

**Comment by divega** [comment]
@lakeman would you be able to test this in 3.0? We believe it is fixed.

**Comment by lakeman** [comment]
Confirmed, that appears to work in 3.0-rc1. Any chance of a fix to 2.2?

**Comment by smitpatel** [comment]
@lakeman - Due to design limitations, this cannot be fixed for 2.2 release.

**Comment by smitpatel** [comment]
Test `Can_read_back_mapped_enum_from_collection_first_or_default` covers this scenario.
